### PR TITLE
[Paddle Backend] Add water tensor dipole example

### PR DESCRIPTION
**Pull request by xusuyong**

### CODE REVIEW

File: deepmd/fit/dipole.py [hunk 1]
```diff
@@ -49,8 +65,9 @@ def __init__(
         precision: str = "default",
         uniform_seed: bool = False,
     ) -> None:
+        super().__init__(name_scope="DipoleFittingSeA")
         """Constructor."""
-        self.ntypes = descrpt.get_ntypes()
+        self.ntypes = descrpt.get_ntypes()  # 2
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 末尾注释可以去掉

File: deepmd/fit/dipole.py [hunk 2]
```diff
@@ -79,79 +146,106 @@ def get_out_size(self) -> int:
...
+            [inputs.shape[0], start_index + natoms, inputs.shape[2]],
+        )
+        inputs_i = paddle.reshape(inputs_i, [-1, self.dim_descrpt])
+        rot_mat_i = paddle.slice(
+            rot_mat,
+            [0, 1, 2],
+            [0, start_index, 0],
+            [rot_mat.shape[0], start_index + natoms, rot_mat.shape[2]],
+        )
+        # paddle.slice(rot_mat, [0, start_index, 0], [-1, natoms, -1])
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 如果注释没有用的话可以删掉

File: deepmd/fit/dipole.py [hunk 3]
```diff
@@ -79,79 +146,106 @@ def get_out_size(self) -> int:
...
+            #         self.n_neuron[ii],
+            #         name="layer_" + str(ii) + suffix,
+            #         reuse=reuse,
+            #         seed=self.seed,
+            #         activation_fn=self.fitting_activation_fn,
+            #         precision=self.fitting_precision,
+            #         uniform_seed=self.uniform_seed,
+            #         initial_variables=self.fitting_net_variables,
+            #         mixed_prec=self.mixed_prec,
+            #     )
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 同上，调试用的注释可以删掉

File: deepmd/fit/dipole.py [hunk 4]
```diff
@@ -79,79 +146,106 @@ def get_out_size(self) -> int:
...
             if (not self.uniform_seed) and (self.seed is not None):
                 self.seed += self.seed_shift
-        # (nframes x natoms) x naxis
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py:119]: 原来的注释不要删除

File: deepmd/fit/dipole.py [hunk 5]
```diff
@@ -79,79 +146,106 @@ def get_out_size(self) -> int:
...
+        #     activation_fn=None,
+        #     name="final_layer" + suffix,
+        #     reuse=reuse,
+        #     seed=self.seed,
+        #     precision=self.fitting_precision,
+        #     uniform_seed=self.uniform_seed,
+        #     initial_variables=self.fitting_net_variables,
+        #     mixed_prec=self.mixed_prec,
+        #     final_layer=True,
+        # )
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 同上

File: deepmd/fit/dipole.py [hunk 6]
```diff
@@ -79,79 +146,106 @@ def get_out_size(self) -> int:
...
-            name="final_layer" + suffix,
-            reuse=reuse,
-            seed=self.seed,
-            precision=self.fitting_precision,
-            uniform_seed=self.uniform_seed,
-            initial_variables=self.fitting_net_variables,
-            mixed_prec=self.mixed_prec,
-            final_layer=True,
         )
+        # # (nframes x natoms) x naxis
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 同上，原有注释尽量不要改动

File: deepmd/fit/dipole.py [hunk 7]
```diff
@@ -79,79 +146,106 @@ def get_out_size(self) -> int:
...
+        # )
         if (not self.uniform_seed) and (self.seed is not None):
             self.seed += self.seed_shift
         # (nframes x natoms) x 1 * naxis
-        final_layer = tf.reshape(
-            final_layer, [tf.shape(inputs)[0] * natoms, 1, self.dim_rot_mat_1]
-        )
+        final_layer = paddle.reshape(
+            final_layer, [paddle.shape(inputs)[0] * natoms, 1, self.dim_rot_mat_1]
+        )  # natoms=64, self.dim_rot_mat_1=100
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 删除调试注释

File: deepmd/fit/dipole.py [hunk 8]
```diff
@@ -79,79 +146,106 @@ def get_out_size(self) -> int:
...
         # (nframes x natoms) x 1 x 3(coord)
-        final_layer = tf.matmul(final_layer, rot_mat_i)
+        final_layer = paddle.matmul(final_layer, rot_mat_i)
         # nframes x natoms x 3
-        final_layer = tf.reshape(final_layer, [tf.shape(inputs)[0], natoms, 3])
-        return final_layer
+        final_layer = paddle.reshape(final_layer, [paddle.shape(inputs)[0], natoms, 3])
+        return final_layer  # [1, 64, 3]
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 删除调试注释

File: deepmd/fit/dipole.py [hunk 9]
```diff
@@ -183,22 +277,25 @@ def build(
...
-        inputs = tf.reshape(input_d, [-1, natoms[0], self.dim_descrpt])
-        rot_mat = tf.reshape(rot_mat, [-1, natoms[0], self.dim_rot_mat])
+        inputs = paddle.reshape(input_d, [-1, natoms[0], self.dim_descrpt])
+        rot_mat = paddle.reshape(rot_mat, [-1, natoms[0], self.dim_rot_mat])
 
         if type_embedding is not None:
-            nloc_mask = tf.reshape(
-                tf.tile(tf.repeat(self.sel_mask, natoms[2:]), [nframes]), [nframes, -1]
+            nloc_mask = paddle.reshape(
+                paddle.tile(paddle.repeat(self.sel_mask, natoms[2:]), [nframes]),
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 没有`paddle.repeat`这个API，对应的应该是`paddle.repeat_interleave`这个API

File: deepmd/fit/dipole.py [hunk 10]
```diff
@@ -183,22 +277,25 @@ def build(
...
+            self.atype_nloc_masked = paddle.reshape(
+                paddle.slice(atype_nall, [0, 0], [-1, natoms[0]])[nloc_mask], [-1]
             )  ## lammps will make error
-            self.nloc_masked = tf.shape(
-                tf.reshape(self.atype_nloc_masked, [nframes, -1])
+            self.nloc_masked = paddle.shape(
+                paddle.reshape(self.atype_nloc_masked, [nframes, -1])
             )[1]
-            atype_embed = tf.nn.embedding_lookup(type_embedding, self.atype_nloc_masked)
+            atype_embed = paddle.nn.embedding_lookup(
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 所有`paddle.nn.xxx`改为`nn.xxx`，因为文件开头已经导入了`nn`

File: deepmd/fit/dipole.py [hunk 11]
```diff
@@ -207,50 +304,57 @@ def build(
         if atype_embed is None:
             count = 0
             outs_list = []
-            for type_i in range(self.ntypes):
+            for type_i in range(self.ntypes):  # 2
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 删除调试代码

File: deepmd/fit/dipole.py [hunk 12]
```diff
@@ -207,50 +304,57 @@ def build(
...
                 if type_i not in self.sel_type:
                     start_index += natoms[2 + type_i]
-                    continue
+                    continue  # sel_type是0，所以就循环了一次
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 删除调试代码

File: deepmd/fit/dipole.py [hunk 13]
```diff
@@ -207,50 +304,57 @@ def build(
...
                     suffix="_type_" + str(type_i) + suffix,
                     reuse=reuse,
+                    type_i=type_i,
                 )
                 start_index += natoms[2 + type_i]
                 # concat the results
                 outs_list.append(final_layer)
                 count += 1
-            outs = tf.concat(outs_list, axis=1)
+
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 删除多余空行

File: deepmd/fit/dipole.py [hunk 14]
```diff
@@ -207,50 +304,57 @@ def build(
...
+            inputs = paddle.reshape(
+                inputs, [nframes, self.nloc_masked, self.dim_descrpt]
+            )
+            rot_mat = paddle.reshape(
                 rot_mat, [nframes, self.nloc_masked, self.dim_rot_mat_1 * 3]
             )
             final_layer = self._build_lower(
                 0, self.nloc_masked, inputs, rot_mat, suffix=suffix, reuse=reuse
             )
+
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 删除多余空行

File: deepmd/fit/dipole.py [hunk 15]
```diff
@@ -207,50 +304,57 @@ def build(
...
             # nframes x natoms x 3
-            outs = tf.reshape(final_layer, [nframes, self.nloc_masked, 3])
+            outs = paddle.reshape(final_layer, [nframes, self.nloc_masked, 3])
 
-        tf.summary.histogram("fitting_net_output", outs)
-        return tf.reshape(outs, [-1])
+        # paddle.summary.histogram("fitting_net_output", outs)
+        return paddle.reshape(outs, [-1])
         # return tf.reshape(outs, [tf.shape(inputs)[0] * natoms[0] * 3 // 3])
```
Comment (HydrogenSulfate) [inline deepmd/fit/dipole.py]: 删除tf代码

File: deepmd/infer/deep_eval.py [hunk 16]
```diff
@@ -77,7 +80,9 @@ def __init__(
         default_tf_graph: bool = False,
         auto_batch_size: Union[bool, int, AutoBatchSize] = False,
     ):
-        jdata = j_loader("input.json")
+        jdata = j_loader(
+            "input.json" if os.path.isfile("input.json") else "dipole_input.json"
```
Comment (HydrogenSulfate) [inline deepmd/infer/deep_eval.py]: 用os.path.exists判断文件是否存在

File: deepmd/infer/deep_eval.py [hunk 17]
```diff
@@ -140,7 +145,12 @@ def __init__(
             if fitting_type == "ener":
                 fitting_param["spin"] = spin
                 fitting_param.pop("type", None)
-            fitting = ener.EnerFitting(**fitting_param)
+                fitting = ener.EnerFitting(**fitting_param)
+            elif fitting_type == "dipole":
+                fitting_param.pop("type", None)
+                fitting = dipole.DipoleFittingSeA(**fitting_param)
+            else:
+                pass
```
Comment (HydrogenSulfate) [inline deepmd/infer/deep_eval.py]: `pass` --> `raise NotImplementedError()`

File: deepmd/infer/deep_eval.py [hunk 18]
```diff
@@ -352,7 +361,8 @@ def __init__(
     @property
     @lru_cache(maxsize=None)
     def model_type(self) -> str:
-        return "ener"
+
+        return self.model.model_type
```
Comment (HydrogenSulfate) [inline deepmd/infer/deep_eval.py]: 删除无用空行

File: deepmd/entrypoints/test.py [hunk 19]
```diff
@@ -890,12 +890,12 @@ def test_dipole(
...
         must=True,
         high_prec=False,
-        type_sel=dp.get_sel_type(),
+        type_sel=dp.tselt,
     )
     test_data = data.get_test()
     dipole, numb_test, atype = run_test(dp, test_data, numb_test, data)
 
-    sel_type = dp.get_sel_type()
+    sel_type = dp.tselt
```
Comment (HydrogenSulfate) [inline deepmd/entrypoints/test.py]: `DeepTensor.get_sel_type`返回的就是tselt，这里不需要改成固定的dp.tselt吧，因为其他DeepXXX类不一定有这个属性，他们是通过get_sel_type方法统一的

File: deepmd/loss/tensor.py [hunk 20]
```diff
@@ -91,53 +93,101 @@ def build(self, learning_rate, natoms, model_dict, label_dict, suffix):
...
+            model_inputs["natoms_vec"],
+            model_inputs["box"],
+            model_inputs["default_mesh"],
+            model_inputs,
+            suffix="",
+            reuse=False,
+        )
+
+        l2_l, l2_more = self.compute_loss(
+            # 0.0, natoms, model_dict, batch_data
```
Comment (HydrogenSulfate) [inline deepmd/loss/tensor.py]: 删除注释

File: deepmd/model/tensor.py [hunk 21]
```diff
@@ -111,63 +113,70 @@ def build(
...
+            # ckpt_meta=ckpt_meta,
             suffix=suffix,
             reuse=reuse,
         )
 
         rot_mat = self.descrpt.get_rot_mat()
-        rot_mat = tf.identity(rot_mat, name="o_rot_mat" + suffix)
+        rot_mat = paddle.clone(rot_mat, name="o_rot_mat" + suffix)
+        # rot_mat = paddle.fluid.layers.assign(rot_mat, name="o_rot_mat" + suffix)
+        # rot_mat = paddle.tensor.clone(rot_mat, name="o_rot_mat" + suffix)
```
Comment (HydrogenSulfate) [inline deepmd/model/tensor.py]: 删除注释

File: deepmd/train/trainer.py [hunk 22]
```diff
@@ -136,7 +137,12 @@ def _init_param(self, jdata):
             if fitting_type == "ener":
                 fitting_param["spin"] = self.spin
                 fitting_param.pop("type")
-            self.fitting = ener.EnerFitting(**fitting_param)
+                self.fitting = ener.EnerFitting(**fitting_param)
+            elif fitting_type == "dipole":
+                fitting_param.pop("type")
+                self.fitting = dipole.DipoleFittingSeA(**fitting_param)
+            else:
+                pass
```
Comment (HydrogenSulfate) [inline deepmd/train/trainer.py]: pass --> `raise NotImplementedError`

File: deepmd/train/trainer.py [hunk 23]
```diff
@@ -798,6 +804,7 @@ def train(self, train_data=None, valid_data=None, stop_batch: int = 10):
         cur_batch = self.global_step
         is_first_step = True
         self.cur_batch = cur_batch
+
```
Comment (HydrogenSulfate) [inline deepmd/train/trainer.py]: 删除空行

**Review by HydrogenSulfate — COMMENTED** [review]
辛苦，没有什么大的问题，小问题修改一下就行

**Review by HydrogenSulfate — APPROVED** [review]
LGTM
